<a href="https://colab.research.google.com/github/Sosswood/Machine-Learning/blob/main/Individual_Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

Import data and add column headers as per the "Read Me" file.  sep='\t' is used because the data is tab separated, rather than comma or pipe.

In [4]:
column_names = ['user_id', 'movie_id', 'rating', 'timestamp']

ratings = pd.read_csv(
    'u.data',
    sep='\t',
    names=column_names
)

Review the data

In [5]:
ratings.head()

,user_id,movie_id,rating,timestamp
0,196,242,3,881250949
1,186,302,3,891717742
2,22,377,1,878887116
3,244,51,2,880606923
4,166,346,1,886397596


Check the size of the data

In [6]:
ratings.shape

(100000, 4)

There are 100,000 items.

Now look into the data content.

In [7]:
ratings.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user_id    100000 non-null  int64
 1   movie_id   100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


There are no missing entries.  All items are integers.

Next, I will look at the overall content of the data.

In [8]:
ratings.describe()

,user_id,movie_id,rating,timestamp
count,100000.00000,100000.000000,100000.000000,1.000000e+05
mean,462.48475,425.530130,3.529860,8.835289e+08
std,266.61442,330.798356,1.125674,5.343856e+06
min,1.00000,1.000000,1.000000,8.747247e+08
25%,254.00000,175.000000,3.000000,8.794487e+08
50%,447.00000,322.000000,4.000000,8.828269e+08
75%,682.00000,631.000000,4.000000,8.882600e+08
max,943.00000,1682.000000,5.000000,8.932866e+08


We can see that the ratings run from 1 - 5.  The mean rating is 3.53 - meaning that most ratings are above the middle rating of 2.5.    The other data is less relevant here as these are ID's, rather than statistical data.

I can gather more meaningful information from this data by changing my questions:

In [9]:
print("Number of ratings:", len(ratings))
print("Number of users:", ratings['user_id'].nunique())
print("Number of movies:", ratings['movie_id'].nunique())
print("Minimum rating:", ratings['rating'].min())
print("Maximum rating:", ratings['rating'].max())

Number of ratings: 100000
Number of users: 943
Number of movies: 1682
Minimum rating: 1
Maximum rating: 5


Now I can see that some users rated more than one film as 943 users rated 100,000 films.  Most films will have received more than one rating as the 100k reviews relate to only 1.6k individual films.

I'm now going to incorporate the names of the films, by adding in the information from u.item, which translates each of the film IDs into a film title and will generally make the data simpler to understand.

In [10]:
movie_columns = [
    'movie_id',
    'movie_title',
    'release_date',
    'video_release_date',
    'IMDb_URL'
]

movies = pd.read_csv(
    'u.item',
    sep='|',
    encoding='latin-1',
    usecols=range(5),
    names=movie_columns
)

movies.head()

,movie_id,movie_title,release_date,video_release_date,IMDb_URL
0,1,Toy Story (1995),01-Jan-1995,NaN,http://us.imdb.com/M/title-exact?Toy%20Story%2...
1,2,GoldenEye (1995),01-Jan-1995,NaN,http://us.imdb.com/M/title-exact?GoldenEye%20(...
2,3,Four Rooms (1995),01-Jan-1995,NaN,http://us.imdb.com/M/title-exact?Four%20Rooms%...
3,4,Get Shorty (1995),01-Jan-1995,NaN,http://us.imdb.com/M/title-exact?Get%20Shorty%...
4,5,Copycat (1995),01-Jan-1995,NaN,http://us.imdb.com/M/title-exact?Copycat%20(1995)


Why is the video release date showing as NaN (Not a Number) in the visible data?  Check the data to see why.

In [11]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1682 entries, 0 to 1681
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   movie_id            1682 non-null   int64  
 1   movie_title         1682 non-null   object 
 2   release_date        1681 non-null   object 
 3   video_release_date  0 non-null      float64
 4   IMDb_URL            1679 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 65.8+ KB


None of the films have a video release date.  We will not use this column going forwards.

Next, merge the data.

In [12]:
df = ratings.merge(
    movies[['movie_id', 'movie_title']],
    on='movie_id',
    how='left'
)

df.head()

,user_id,movie_id,rating,timestamp,movie_title
0,196,242,3,881250949,Kolya (1996)
1,186,302,3,891717742,L.A. Confidential (1997)
2,22,377,1,878887116,Heavyweights (1994)
3,244,51,2,880606923,Legends of the Fall (1994)
4,166,346,1,886397596,Jackie Brown (1997)


Check that all rows of data are still present following the merge:

In [13]:
print("Number of rows:", len(df))
print("Number of users:", df['user_id'].nunique())
print("Number of movies:", df['movie_id'].nunique())

df.head(10)

Number of rows: 100000
Number of users: 943
Number of movies: 1682


,user_id,movie_id,rating,timestamp,movie_title
0,196,242,3,881250949,Kolya (1996)
1,186,302,3,891717742,L.A. Confidential (1997)
2,22,377,1,878887116,Heavyweights (1994)
3,244,51,2,880606923,Legends of the Fall (1994)
4,166,346,1,886397596,Jackie Brown (1997)
5,298,474,4,884182806,Dr. Strangelove or: How I Learned to Stop Worr...
6,115,265,2,881171488,"Hunt for Red October, The (1990)"
7,253,465,5,891628467,"Jungle Book, The (1994)"
8,305,451,3,886324817,Grease (1978)
9,6,86,3,883603013,"Remains of the Day, The (1993)"


The number of rows, number of users and number of films remains the same as before:  
Number of rows: 100000
Number of users: 943
Number of movies: 1682

Ensure that both ratings and film data are kept.

In [14]:
ratings
movies
df

,user_id,movie_id,rating,timestamp,movie_title
0,196,242,3,881250949,Kolya (1996)
1,186,302,3,891717742,L.A. Confidential (1997)
2,22,377,1,878887116,Heavyweights (1994)
3,244,51,2,880606923,Legends of the Fall (1994)
4,166,346,1,886397596,Jackie Brown (1997)
...,...,...,...,...,...
99995,880,476,3,880175444,"First Wives Club, The (1996)"
99996,716,204,5,879795543,Back to the Future (1985)
99997,276,1090,1,874795795,Sliver (1993)
99998,13,225,2,882399156,101 Dalmatians (1996)
